Improve README (#16)
ethereum/act: README.md

Act
===

This project is an effort by several groups working on formal methods for
Ethereum smart contracts, aiming at creating a simple but effective language
to write formal specification.
It extends on the previous
[Act](https://github.com/dapphub/klab/blob/master/acts.md) project.

Some general features it hopes to achieve are:
- Simple structure and logic.
- Support writing properties on a high level (closer to Solidity/Vyper) as
  well as on the machine level.
- Cross-tool communication: Different tools might be able to generate
  different knowledge and use the specification language to exchange
  information.
- No attachment to a particular logic. If a tool wishes to support certain
  expressions, extensions can be written.

Everyone is encouraged to join/start discussions on issues and pull requests.

Language
========

> Acts are based on behaviours from the specified smart contract. Take the

Reviewer: ```suggestion
Act specifications are functional descriptions of the behaviour of a smart contract. Take the
```

following toy state machine written in Solidity as an example:

```sol
contract StateMachine {
    uint x;
  
    function f() public {      
        if (x == 0)
            x = 1;             
    }

    function g() public {
        if (x == 1)
            x = 0;
    }
}
```

The initial state is `x = 0`, and `x` can toggle between 1 and 0 by calling
functions `f` and `g`, respectively.

A simple specification for the program above could be the following Act:

```act
behaviour init of StateMachine
interface constructor()

creates
	uint256 x := 0

> invariants

Reviewer: ooh, I like how you put the invariants in the `contructor`!

Reviewer: This might be a better idea than having loose hanging `property/invariant` sections.

Author: Yea they were loose before but there was kind of this idea of a super section `contract`. Since this doesn't exist I wasn't sure where to put them, so just put it there. I think it works too.

	x <= 1

behaviour f of StateMachine
interface f()

> case x == 0:

Reviewer: missing `storage` keyword everywhere in this spec

Reviewer: and the syntax of storage updates is `=>`

	x := 1

case _:
	noop

ensures

> 	(x == 0) || (x == 1)

Reviewer: ```suggestion
	(x == 0) or (x == 1)
```

behaviour g of StateMachine
interface g()

case x == 1:
	x := 0

case _:
	noop

ensures
	(x == 1) || (x == 0)
```

A `case` inside a `behaviour` specifies how storage changes, and is the
lowest level mechanism in the specification. Functions post conditions are
described in the `ensures` section. Contract invariants are high level
properties given for the entire contract.

> 

Reviewer: ```suggestion
The `case`s of a `behaviour` specify how storage changes as a result of calling the function, and its return argument, if present. They make up the
lowest level description of the specification. Functions post conditions are
described in the `ensures` section. Contract invariants specify relations between its storage variables that should remain true for the entire lifetime of the contract.

```

More examples can be found in the `examples` directory.

Modular Approach
================

Act is designed to enable modular verification. Instead of checking
properties that might be too hard to prove from the bytecode, such as
contract invariants, we hope that it is easier to do that via multiple easier
steps:

1. Given `behaviour case`, show that `post storage` is implemented by the

>    bytecode.

Reviewer: ```suggestion
1. Given a `behaviour`, show that implementation bytecode results updates storage and return argument as specified.
```

2. Given `behaviour`, prove that the `post condition` holds.
3. Given `post conditions`, show contract invariant properties.
4. Given `(transition system = "CONTRACT")`, show that arbitrary properties

>    hold.

Reviewer: ```suggestion
2. Given a `behaviour`, prove that the post condition described in the `ensures` section holds.
3. Given a set of `behaviour`s, prove contract invariants.
4. Given `(transition system = "CONTRACT")`, show that arbitrary properties
   hold.
```

Note that steps 2, 3 and 4 can be done over Act only, without interaction
with the source code/bytecode. Checking that the code implements the

> behaviour specification correctly should be much easier than proving that

Reviewer: hmm, this depends...

Author: Yea.. any ideas how to rephrase this? Or just remove the second sentence?

Reviewer: We can maybe just remove it for now

        emit Transfer(from, to, value);
        return true;
    }
    function approve(address spender, uint value) public returns (bool) {
        allowance[msg.sender][spender] = value;
        emit Approval(msg.sender, spender, value);
        return true;
    }
}
```

The behaviours of the constructor and the functions `transfer` and
`transferFrom` can be specified as the following Act:

```act
behaviour init of Token
interface constructor(string _symbol, string _name, string _version, uint _totalSupply)

creates
    string name         := _name
    string symbol       := _symbol
    uint256 totalSupply := _totalSupply
    mapping(address => uint) balanceOf :=  [CALLER := _totalSupply]
    mapping(address=>mapping(address=>uint)) allowance := []

behaviour transfer of Token
interface transfer(uint256 value, address to)

iff

   CALLVALUE == 0
   value <= balanceOf[CALLER]
   CALLER =/= to => balanceOf[to] + value < 2^256

case CALLER =/= to:

   storage

     balanceOf[CALLER] => balanceOf[CALLER] - value
     balanceOf[to]     => balanceOf[to] + value

   returns 1

case CALLER == to:

   returns 1

behaviour transferFrom of Token

> interface transferFrom(address src, address dst, uint amount)

Reviewer: ```suggestion
interface transferFrom(address src, address dst, uint value)
```

iff

   value <= balanceOf[CALLER]
   src    =/= dst => balanceOf[dst] + value < 2^256
   CALLER =/= src => 0 <= allowance[src][CALLER] - value
   CALLVALUE == 0

case src =/= dst:

   case CALLER == src:

      storage

         balances[src] => balances[src] - value
         balances[dst] => balances[dst] + value

      returns 1

   case CALLER =/= src and allowance[src][CALLER] == 2^256 - 1:

      storage

         balances[src] => balances[src] - value
         balances[dst] => balances[dst] + value

      returns 1

   case _:

      storage

         allowance[src][CALLER] => allowance[src][CALLER] - value
         balances[src]          => balances[src] - value
         balances[dst]          => balances[dst] + value

      returns 1

case src == dst:

   returns 1
```

> Parsing te Act gives us the generated proof obligations:

Reviewer: ```suggestion
Parsing the Act gives us the generated proof obligations:
```

> ```JSON

Reviewer: notice that this json is from a simpler, `mint` spec and not the `transfer`  specified above.

Author: Oh. Do we have the spec somewhere? Or the json for the transfer above?

Reviewer: Not currently, but it's probably the next reachable target

Author: Should I update the act and code to `mint`, or just remove the json?

Reviewer: It's good to show the json I think, so yeah let's show the spec of mint right now
